File: python/algos/implement_str_str.py
def sub_str(haystack: str, needle: str) -> int:
    """Finds index of first occurrence of needle in haystack."""
    if not needle:
        return 0
    index = 0
    while index <= len(haystack) - len(needle):
        i = 0
        while i < len(needle) and haystack[index + i] == needle[i]:
            i += 1

        if i == len(needle):
            return index

        index += 1
    return -1


if __name__ == '__main__':
    assert sub_str("hello", "") == 0
    assert sub_str("hello", "ll") == 2
    assert sub_str("hello", "lls") == -1
    assert sub_str("aaaa", "aaa") == 0

File: rust/src/algos/implement_str_str.rs
pub fn str_str(haystack: String, needle: String) -> i32 {
    if needle.len() == 0 {
        return 0;
    }
    if needle.len() > haystack.len() {
        return -1
    }
    let haystack_chars: Vec<char> = haystack.chars().collect();
    let needle_chars: Vec<char> = needle.chars().collect();
    let mut index = 0;
    while index <= haystack_chars.len() - needle_chars.len() {
        let mut i = 0;

        while i < needle_chars.len() && haystack_chars[index + i] == needle_chars[i] {
            i += 1;
        }
        if i == needle_chars.len() {
            return index as i32
        }
        index += 1
    }
    -1
}
